templates/chat/src/components/WhiteboardModal.tsx

```diff
@@ -0,0 +1,186 @@
+	notifyIfFileNotAllowed,
+	TLComponents,
+	Tldraw,
+	TldrawOptions,
+	TldrawUiButton,
+	TldrawUiRow,
+	TLEditorSnapshot,
+	TLImageShape,
+	useEditor,
+	useToasts,
+	useTranslation,
+} from 'tldraw'
+
+export interface TldrawProviderMetadata {
+	snapshot: TLEditorSnapshot
+	imageName: string
+}
+
+export interface WhiteboardImage {
+	id: string
+	name: string
+	url: string
+	snapshot: TLEditorSnapshot
+	type: string
+	width: number
+	height: number
+}
+
+interface WhiteboardModalProps {
+	initialSnapshot?: TLEditorSnapshot
+	onCancel: () => void
+	onAccept: (image: WhiteboardImage) => void
+	imageId?: string
+	uploadedFile?: File
+	imageName?: string
+}
+
+const options: Partial<TldrawOptions> = {
+	// disable the ability to create new pages:
+	maxPages: 1,
+	// make sure the action shortcuts are always in the top-right menu area, not on the toolbar:
+	actionShortcutsLocation: 'menu',
+	// disable font pre-loading to avoid the ui popping in after the modal appears:
+	maxFontsToLoadBeforeRender: 0,
+}
+
+export function WhiteboardModal({
+	initialSnapshot,
+	onCancel,
+	onAccept,
+	imageId,
+	uploadedFile,
+	imageName,
+}: WhiteboardModalProps) {
+	// components are used to override parts of the tldraw ui. they shouldn't change often, so it's
+	// important that we memoize them or define them outside the tldraw component.
+	const components = useMemo(
+		(): TLComponents => ({
+			// The "SharePanel" is in the top-right of the editor. Here we want it to show our save
+			// and cancel buttons:
+			SharePanel: () => {
+				const editor = useEditor()
+				return (
+					<TldrawUiRow className="whiteboard-actions">
+						<TldrawUiButton type="normal" onClick={onCancel}>
+							Cancel
+						</TldrawUiButton>
+						<TldrawUiButton
+							type="primary"
+							onClick={async () => {
+								// when the user clicks save, we convert the current whiteboard to an image:
+								const image = await editor.toImageDataUrl(editor.getCurrentPageShapes(), {
+									format: 'png',
+								})
+
+								// we also take a snapshot of the editor state, so we can still edit
+								// it if we open it up again later:
+								const snapshot = editor.getSnapshot()
+
+								// we pass the image data and the snapshot to the parent component, so it
+								// can add it to the chat input:
+								onAccept({
+									id: imageId ?? crypto.randomUUID(),
+									name: imageName ?? 'tldraw whiteboard.png',
+									snapshot,
+									type: 'image/png',
+									...image,
+								})
+							}}
+						>
+							{imageId ? 'Save' : 'Add'}
+						</TldrawUiButton>
+					</TldrawUiRow>
+				)
+			},
+		}),
+		[onCancel, onAccept, imageId, imageName]
+	)
+
+	// when the user clicks outside the modal, we close it:
+	const handleOverlayClick = (e: React.MouseEvent) => {
+		if (e.target === e.currentTarget) {
+			onCancel()
+		}
+	}
+
+	return (
+		<div className="modal-overlay" onClick={handleOverlayClick}>
+			<Tldraw
+				components={components}
```

Comment:
something odd, when the modal shows (for an existing whiteboard), it auto-focuses a random(?) shape. why is that? might not be related to your PR